src/utilities/mta/range_proofs.rs

```diff
@@ -0,0 +1,718 @@
+        let h1 = &dlog_statement.g;
+        let h2 = &dlog_statement.ni;
+
+        if self.s1 > FE::q().pow(3) {
+            return false;
+        }
+
+        let z_e_inv = BigInt::mod_inv(&BigInt::mod_pow(&self.z, &self.e, N_tilde), N_tilde);
+        if z_e_inv.is_none() {
+            // z must be invertible, yet the check is done here
+            return false;
+        }
+        let z_e_inv = z_e_inv.unwrap();
+
+        let z_prim = (BigInt::mod_pow(h1, &self.s1, N_tilde)
+            * BigInt::mod_pow(h2, &self.s2, N_tilde)
+            * z_e_inv)
+            % N_tilde;
+
+        let mta_e_inv = BigInt::mod_inv(&BigInt::mod_pow(mta_avc_out, &self.e, NN), NN);
+        if mta_e_inv.is_none() {
+            // mta_avc_out must be invertible, yet the check is done here
+            return false;
+        }
+        let mta_e_inv = mta_e_inv.unwrap();
+
+        let v = (BigInt::mod_pow(a_enc, &self.s1, NN)
+            * BigInt::mod_pow(&self.s, N, NN)
+            * (self.t1.borrow() * N + 1)
+            * mta_e_inv)
+            % NN;
+
+        let t_e_inv = BigInt::mod_inv(&BigInt::mod_pow(&self.t, &self.e, N_tilde), N_tilde);
+        if t_e_inv.is_none() {
+            // t must be invertible, yet the check is done here
+            return false;
+        }
+        let t_e_inv = t_e_inv.unwrap();
+        let w = (BigInt::mod_pow(h1, &self.t1, N_tilde)
+            * BigInt::mod_pow(h2, &self.t2, N_tilde)
+            * t_e_inv)
+            % N_tilde;
+
+        let Gen = alice_ek.n.borrow() + 1;
+        let mut values_to_hash = vec![
+            &alice_ek.n,
+            &Gen,
+            a_enc,
+            mta_avc_out,
+            &self.z,
+            &z_prim,
+            &self.t,
+            &v,
+            &w,
+        ];
+        let e = if check.is_some() {
+            let X_x_coor = check.unwrap().X.x_coor().unwrap();
+            values_to_hash.push(&X_x_coor);
+            let X_y_coor = check.unwrap().X.y_coor().unwrap();
+            values_to_hash.push(&X_y_coor);
+            let u_x_coor = check.unwrap().u.x_coor().unwrap();
+            values_to_hash.push(&u_x_coor);
+            let u_y_coor = check.unwrap().u.y_coor().unwrap();
+            values_to_hash.push(&u_y_coor);
+            HSha256::create_hash(&values_to_hash[..])
+        } else {
+            HSha256::create_hash(&values_to_hash[..])
+        };
+
+        if e != self.e {
+            return false;
+        }
+
+        true
+    }
+
+    pub fn generate(
+        a_encrypted: &BigInt,
+        mta_encrypted: &BigInt,
+        b: &FE,
+        beta_prim: &BigInt,
+        alice_ek: &EncryptionKey,
+        dlog_statement: &DLogStatement,
+        r: &Randomness,
+        check: bool,
+    ) -> (BobProof, Option<GE>) {
+        let round1 = BobZkpRound1::from(
+            alice_ek,
+            dlog_statement,
+            b,
+            beta_prim,
+            a_encrypted,
+            &FE::q(),
```

Comment:
I guess `q` doesn't need to be passed here

Reply:
If I understood you correctly, and you wanted it to be defined inside Bob's round 1, I made it a parameter just so that any potential efforts to make MtA generic over the curve are a tiny bit easier :)